feat: Windows.Devices.Sensors.Pedometer

MartinZikmund opened the pull request:
GitHub Issue (If applicable): #2069

## PR Type

- Feature

## What is the current behavior?

`Pedometer` is not supported.

## What is the new behavior?

`Pedometer` is supported for active step-counting :-) .

This PR also unifies the default `ReportInterval` of other sensors on Android (previously it was not adjustable)


## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](https://github.com/unoplatform/uno/blob/master/README.md#uno-features)
- [x] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Updated the [Release Notes](https://github.com/unoplatform/uno/tree/master/doc/ReleaseNotes)
- [x] Associated with an issue (GitHub or internal)

carldebilly commented:
Awesome.

carldebilly commented on src/Uno.UWP/Generated/3.0.0.0/Windows.Devices.Sensors/PedometerReading.cs:
```diff
@@ -2,51 +2,15 @@
 namespace Windows.Devices.Sensors
 {
-	#if __ANDROID__ || __IOS__ || NET461 || __WASM__ || __MACOS__
+	#if false || false || false || false || false
```
Mayby it should still `NotImplemented` for unsupported platforms :-).

carldebilly commented:
/azp run

carldebilly commented on src/SamplesApp/UITests.Shared/Windows_Devices/PedometerTests.xaml.cs:
```diff
@@ -0,0 +1,144 @@
+
+		public bool IsAvailable => _pedometer != null;
+
+		private async void GetPedometerAsync()
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Return 'Task' instead.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826529)

jeromelaban commented on doc/ReleaseNotes/_ReleaseNotes.md:
```diff
@@ -2,9 +2,11 @@
 - Support for `AnalyticsInfo` on macOS
 - Support for `TryEnterFullScreenMode` and `ExitFullScreenMode` on WebAssembly
 - Support for `MessageDialog` on macOS
+>>>>>>> upstream/master
```
The merge was incorrect.

jeromelaban commented on doc/articles/features/windows-devices-sensors.md:
```diff
@@ -45,3 +45,28 @@ On Android, when both `ReadingChanged` and `Shaken` events are attached and the
+
+#### Android
+
+On Android, the first reading returns the cumulative number of steps since the device was first booted up. I have noticed the sensor does not correctly respect the requested reporting interval, so we implemented this manually to make sure the `ReadingChanged` events are triggered only after the `ReportInterval` elapses.
```
```suggestion
On Android, the first reading returns the cumulative number of steps since the device was first booted up. The sensor may not correctly respect the requested reporting interval, so the implementation does this manually to make sure the `ReadingChanged` events are triggered only after the `ReportInterval` elapses.
```

jeromelaban commented on src/Uno.UWP/Devices/Sensors/Pedometer.cs:
```diff
@@ -0,0 +1,67 @@
+					_instance = TryCreateInstance();
+					_initializationAttempted = true;
+				}
+				return Task.FromResult(_instance).AsAsyncOperation();
```
Is there a reason why the initalization is not completely async ? Is this operation fast ?

MartinZikmund replied:
As the other platforms don't have asynchronous retrieval of the sensor, I didn't consider it originally. It is fast, but I could wrap the initialization in a `Task.Run` if that would be more appropriate.

MartinZikmund replied:
Done :-) .

carldebilly commented on src/SamplesApp/UITests.Shared/Windows_Devices/PedometerTests.xaml.cs:
```diff
@@ -0,0 +1,152 @@
+
+		public bool IsAvailable => _pedometer != null;
+
+		private async void GetPedometerAsync()
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Return 'Task' instead.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826529)

dr1rrb replied:
Which is not supported by the `GetOrCreateCommand`. We could add it, but we would face the question of how to handle multiple parallel executions of the command, how to handle errors, etc. ... for tests.

MartinZikmund replied:
I would think this could be ok for the sample.

davidjohnoliver commented on src/Uno.UWP/Devices/Sensors/Helpers/SensorHelpers.iOS.cs:
```diff
@@ -6,12 +6,49 @@ namespace Uno.Devices.Sensors.Helpers
 			return new DateTimeOffset(date);
 		}
+
+		public static DateTimeOffset NSDateToDateTimeOffset(NSDate nsDate)
```
There are already similar methods in `NSDateExtensions`, do they work for your purpose?

MartinZikmund replied:
Unfortunately they are in `Uno.UI`, so I can't use those...

davidjohnoliver replied:
Ah. Do they work, though? If so just move them. (Or move the implementation and redirect the `Uno.UI` methods for binary compatibility)

MartinZikmund replied:
Just realized they are actually a bit different - the `NSDateExtensions` work with `DateTime` while Sensors require `DateTimeOffset` instead, so it is not exactly the same and would require additional methods anyway.

davidjohnoliver commented on src/Uno.UWP/Devices/Sensors/Pedometer.Android.cs:
```diff
@@ -0,0 +1,105 @@
+						{
+							//restart reading to apply interval
+							StartReading();
+							StopReading();
```
Order seems inverted here

MartinZikmund replied:
Thanks! Fixed.

nventive-devops commented on src/SamplesApp/UITests.Shared/Windows_Devices/PedometerTests.xaml.cs:
```diff
@@ -0,0 +1,152 @@
+
+		public bool IsAvailable => _pedometer != null;
+
+		private async void GetPedometerAsync()
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Return 'Task' instead.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826529)

nventive-devops replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Return 'Task' instead.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826529)

nventive-devops replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Return 'Task' instead.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826529)

nventive-devops replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Return 'Task' instead.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826529)

nventive-devops replied:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Return 'Task' instead.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826529)

MartinZikmund commented:
@jeromelaban I will need to update my existing Pull Requests to support new conventional commits requirement, how do I do that please? When I try to merge upstream and push, I get:

```
! [remote rejected]     dev/mazi/pedometer -> dev/mazi/pedometer (refusing to allow an OAuth App to create or update workflow `.github/workflows/conventional-commits.yml` without `workflow` scope) 
```

Same when I rebase and try force push. Is there a way to fix that or will I have to create a new branch and PR?

Also - for existing commits - do I mark each progress commit as "feat:"? For example 
feat: Implemented Pedometer on iOS, feat: Implemented Pedometer on Android, etc.? I am not sure how it will look in the new changelog, wouldn't want to clutter it with each smaller change commit. Is that what chore: is for?

jeromelaban commented:
/azp run
